modules/pam_setquota/pam_setquota.c

```
@@ -0,0 +1,273 @@
+      size_t mnt_len = strlen(mnt->mnt_dir);
+      const char *s;
+      if (mnt_len > match_size &&
+         (s = pam_str_skip_prefix_len(pwd->pw_dir, mnt->mnt_dir, mnt_len)) != NULL &&
+         (s[0] == '\0' || s[0] == '/')) {
+        free(mntdevice);
+        if ((mntdevice = strdup(mnt->mnt_fsname)) == NULL) {
+          pam_syslog(pamh, LOG_CRIT,  "Memory allocation error");
+          endmntent(fp);
+          return PAM_PERM_DENIED;
+        }
+        match_size = mnt_len;
+      }
+    } else if ((strcmp(param.fs, mnt->mnt_dir) == 0) ||
+               (strcmp(param.fs, mnt->mnt_fsname) == 0)) {
+        free(mntdevice);
+        if ((mntdevice = strdup(mnt->mnt_fsname)) == NULL) {
+          pam_syslog(pamh, LOG_CRIT,  "Memory allocation error");
+          endmntent(fp);
+          return PAM_PERM_DENIED;
+        }
+    }
+  }
+
+  endmntent(fp);
+
+  if (mntdevice == NULL) {
+    pam_syslog(pamh, LOG_ERR, "Filesystem not found");
+    return PAM_PERM_DENIED;
+  }
+
+  /* Get limits */
+  if (quotactl(QCMD(Q_GETQUOTA, USRQUOTA), mntdevice, pwd->pw_uid,
+               (void *)&ndqblk) == -1) {
+    pam_syslog(pamh, LOG_ERR, "fail to get limits for user %s : %m",
+               pwd->pw_name);
+    free(mntdevice);
+    return PAM_PERM_DENIED;
+  }
+
+  if (param.debug == 1)
+    debug(pamh, &ndqblk);
+
+  /* Only overwrite if quotas aren't already set or if overwrite is set */
+  if ((ndqblk.dqb_bsoftlimit == 0 && ndqblk.dqb_bhardlimit == 0 &&
+       ndqblk.dqb_isoftlimit == 0 && ndqblk.dqb_ihardlimit == 0) ||
+      param.overwrite == 1) {
+
+    /* Parse new limits */
+    ndqblk.dqb_valid = 0;
+    if (!_pam_parse_dqblk(argc, argv, &ndqblk)) {
+      pam_syslog(pamh, LOG_AUTH | LOG_WARNING, "setting soft and hard limits"
+                                               "separately");
+    }
+
+    /* Set limits */
+    if (quotactl(QCMD(Q_SETQUOTA, USRQUOTA), mntdevice, pwd->pw_uid,
+                 (void *)&ndqblk) == -1) {
```

Is there any reason to call setquota if ndqblk.dqb_valid == 0?

---

You mean if `ndqblk.dqb_valid` remains at 0 after  `_pam_parse_dqblk()`?

---

Yes.

---

Yes, one could check for `ndqblk.dqb_valid != 0` but for this to happen you would need to configure pam_setquota without any limits and why would anyone do this?

---

By mistake?  This would justify issuing a warning.

---

Fair enough. Something like:

```
    /* No limits defined at all in configuration?*/
    if (ndqblk.dqb_valid == 0) {
      pam_syslog(pamh, LOG_AUTH | LOG_WARNING, "no limits defined in configuration");
    } 
```  
I am unsure if the module should exit in that `if` and if, then with what exit code, or just let the code run along the normal path and exit at the end.

---

After sleeping over this: I don't think the module should throw an error here. Sure, this is unusual, but since no harm is done and the module then is just a NO-OP, it should just warn and then exit normally.